db/src/tx_observer.rs

```diff
@@ -0,0 +1,242 @@
+    observers: Arc<IndexMap<String, Arc<TxObserver>>>,
+    transactions: BTreeMap<Entid, AttributeSet>,
+    executor: Option<Sender<Box<Command + Send>>>,
+}
+
+impl TxObservationService {
+    pub fn new() -> Self {
+        TxObservationService {
+            observers: Arc::new(IndexMap::new()),
+            transactions: Default::default(),
+            executor: None,
+        }
+    }
+
+    // For testing purposes
+    pub fn is_registered(&self, key: &String) -> bool {
+        self.observers.contains_key(key)
+    }
+
+    pub fn register(&mut self, key: String, observer: Arc<TxObserver>) {
+        Arc::make_mut(&mut self.observers).insert(key, observer);
+    }
+
+    pub fn deregister(&mut self, key: &String) {
+        Arc::make_mut(&mut self.observers).remove(key);
+    }
+
+    pub fn has_observers(&self) -> bool {
+        !self.observers.is_empty()
+    }
+
+    pub fn add_transaction(&mut self, tx_id: Entid, attributes: AttributeSet) {
+        self.transactions.insert(tx_id, attributes);
+    }
+
+    pub fn transaction_did_commit(&mut self, txids: &AccumulatedTxids) {
+        // collect the changesets relating to this commit
+        let reports: BTreeMap<Entid, AttributeSet> = txids.into_iter().filter_map(|tx_id| {
+                                        self.transactions.remove(&tx_id).map_or(None, |changeset| Some((tx_id, changeset)))
+                                        })
+                                        .fold(BTreeMap::new(), |mut map, (tx_id, changeset)| {
+                                            map.insert(*tx_id, changeset);
+                                            map
+                                        });
+
+        let executor = self.executor.get_or_insert_with(||{
+            let (tx, rx): (Sender<Box<Command + Send>>, Receiver<Box<Command + Send>>) = channel();
+            let mut worker = CommandExecutor::new(rx);
+
+            thread::spawn(move || {
+                worker.main();
+            });
+
+            tx
+        });
+
+        let cmd = Box::new(TxCommand::new(&self.observers, reports));
+        executor.send(cmd).unwrap();
+    }
+}
+
+impl Drop for TxObservationService {
+    fn drop(&mut self) {
+        self.executor = None;
+    }
+}
+
+pub struct InProgressObserverTransactWatcher<'a> {
+    collected_datoms: AttributeSet,
+    observer_service: &'a Mutex<TxObservationService>,
+    active: bool
+}
+
+impl<'a> InProgressObserverTransactWatcher<'a> {
+    pub fn new(observer_service: &'a Mutex<TxObservationService>) -> InProgressObserverTransactWatcher {
+        let mut w = InProgressObserverTransactWatcher {
+            collected_datoms: Default::default(),
+            observer_service,
+            active: true
+        };
+
+        w.active = observer_service.lock().unwrap().has_observers();
+        w
+    }
+}
+
+impl<'a> TransactWatcher for InProgressObserverTransactWatcher<'a> {
+    type Result = ();
+
+    fn tx_id(&mut self) -> Option<Entid> {
```

Comment:
I don't think `tx_id` belongs in `TransactWatcher`. By all means put it in `impl`s for specific types, but it's not a meaningful part of the trait, and all of these `None`s show that.